src/index.js

```diff
@@ -97,60 +75,54 @@ class IpfsRepo {
    */
   open (callback) {
     if (!this.closed) {
-      return callback(new Error('repo is already open'))
+      callback(new Error('repo is already open'))
+      return // early
     }
     log('opening at: %s', this.path)
 
     // check if the repo is already initialized
     waterfall([
-      (cb) => this._fsStore.open((err) => {
-        if (err && err.message === 'Already open') {
-          return cb()
-        }
-        cb(err)
-      }),
+      (cb) => this.root.open(ignoringAlreadyOpened(cb)),
       (cb) => this._isInitialized(cb),
       (cb) => this._locker.lock(this.path, cb),
       (lck, cb) => {
         log('aquired repo.lock')
         this.lockfile = lck
-
-        log('creating flatfs')
-        const FsStore = this.options.fs
-        const s = new FsStore(path.join(this.path, flatfsDirectory), Object.assign({}, this.options.fsOptions))
-
-        if (this.options.sharding) {
-          const shard = new core.shard.NextToLast(2)
-          ShardingStore.createOrOpen(s, shard, cb)
-        } else {
-          cb(null, s)
-        }
+        cb()
       },
-      (flatfs, cb) => {
-        log('Flatfs store opened')
-        this.store = new MountStore([{
-          prefix: new Key(flatfsDirectory),
-          datastore: flatfs
-        }, {
-          prefix: new Key('/'),
-          datastore: new LevelStore(path.join(this.path, levelDirectory), {
-            db: this.options.level
-          })
-        }])
-
-        this.blockstore = blockstore(this)
+      (cb) => {
+        log('creating datastore')
+        this.datastore = backends.create('datastore', path.join(this.path, 'datastore'), this.options)
+        log('creating blocks')
+        const blocksBaseStore = backends.create('blocks', path.join(this.path, 'blocks'), this.options)
+        blockstore(
+          blocksBaseStore,
+          this.datastore,
```

Comment:
ah, wait.

this.datastore is /datastore? If so, why does /blocks need it?
if this.datastore is just /, then it makes sense to call it rootDatastore